### ISSUE TITLE
Duplicated LSP servers with Pyright since b0bd429.

### ISSUE THREAD

**Issue by tysonliddell** [neovim/nvim-lspconfig#2360]
### Description

Each new buffer spawns a new pyright LSP server since the logic to determine LSP workspace support was changed with b0bd429.

### Neovim version

NVIM v0.8.1
Build type: Release
LuaJIT 2.1.0-beta3

### Nvim-lspconfig version

a63d2c7

### Operating system and version

macOS 12.6

### Affected language servers

pyright

### Steps to reproduce

1. Create a new directory with empty files `pyproject.toml`, `foo.py` and `bar.py`.
2. `cd` into the directory and run `nvim foo.py`.
3. Run `:e bar.py` inside `nvim`.
4. Observe `:LspInfo` shows two language servers running, one for each buffer.

### Actual behavior

`:LspInfo` shows two language servers running, one for each buffer.
```
 Language client log: /Users/tyson/.local/state/nvim/lsp.log
 Detected filetype:   python
 
 1 client(s) attached to this buffer: 
 
 Client: pyright (id: 2, bufnr: [3])
 	filetypes:       python
 	autostart:       true
 	root directory:  /Users/tyson/aaaaaaa
 	cmd:             pyright-langserver --stdio
 
 1 active client(s) not attached to this buffer: 
 
 Client: pyright (id: 1, bufnr: [1])
 	filetypes:       python
 	autostart:       true
 	root directory:  /Users/tyson/aaaaaaa
 	cmd:             pyright-langserver --stdio
 
 Configured servers list: pyright, tsserver, rust_analyzer
```

This breaks the ability to rename a variable across multiple files, as the separate language servers appear to get out of sync. It probably breaks other LSP functionality as well.

### Expected behavior

The correct behaviour occurs when reverting to commit 22c87d6 (tag: v0.1.4).
```
 Language client log: /Users/tyson/.local/state/nvim/lsp.log
 Detected filetype:   python
 
 1 client(s) attached to this buffer: 
 
 Client: pyright (id: 1, bufnr: [1, 3])
 	filetypes:       python
 	autostart:       true
 	root directory:  /Users/tyson/aaaaaaa
 	cmd:             pyright-langserver --stdio
 
 Configured servers list: tsserver, rust_analyzer, pyright
```

### Minimal config

```Lua
-- This config is identical to the current minimal_init.lua.
local on_windows = vim.loop.os_uname().version:match 'Windows'

local function join_paths(...)
  local path_sep = on_windows and '\\' or '/'
  local result = table.concat({ ... }, path_sep)
  return result
end

vim.cmd [[set runtimepath=$VIMRUNTIME]]

local temp_dir = vim.loop.os_getenv 'TEMP' or '/tmp'

vim.cmd('set packpath=' .. join_paths(temp_dir, 'nvim', 'site'))

local package_root = join_paths(temp_dir, 'nvim', 'site', 'pack')
local lspconfig_path = join_paths(package_root, 'test', 'start', 'nvim-lspconfig')

if vim.fn.isdirectory(lspconfig_path) ~= 1 then
  vim.fn.system { 'git', 'clone', 'https://github.com/neovim/nvim-lspconfig', lspconfig_path }
end

vim.lsp.set_log_level 'trace'
require('vim.lsp.log').set_format_func(vim.inspect)
local nvim_lsp = require 'lspconfig'
local on_attach = function(_, bufnr)
  local function buf_set_option(...)
    vim.api.nvim_buf_set_option(bufnr, ...)
  end

  buf_set_option('omnifunc', 'v:lua.vim.lsp.omnifunc')

  -- Mappings.
  local opts = { buffer = bufnr, noremap = true, silent = true }
  vim.keymap.set('n', 'gD', vim.lsp.buf.declaration, opts)
  vim.keymap.set('n', 'gd', vim.lsp.buf.definition, opts)
  vim.keymap.set('n', 'K', vim.lsp.buf.hover, opts)
  vim.keymap.set('n', 'gi', vim.lsp.buf.implementation, opts)
  vim.keymap.set('n', '<C-k>', vim.lsp.buf.signature_help, opts)
  vim.keymap.set('n', '<space>wa', vim.lsp.buf.add_workspace_folder, opts)
  vim.keymap.set('n', '<space>wr', vim.lsp.buf.remove_workspace_folder, opts)
  vim.keymap.set('n', '<space>wl', function()
    print(vim.inspect(vim.lsp.buf.list_workspace_folders()))
  end, opts)
  vim.keymap.set('n', '<space>D', vim.lsp.buf.type_definition, opts)
  vim.keymap.set('n', '<space>rn', vim.lsp.buf.rename, opts)
  vim.keymap.set('n', 'gr', vim.lsp.buf.references, opts)
  vim.keymap.set('n', '<space>e', vim.diagnostic.open_float, opts)
  vim.keymap.set('n', '[d', vim.diagnostic.goto_prev, opts)
  vim.keymap.set('n', ']d', vim.diagnostic.goto_next, opts)
  vim.keymap.set('n', '<space>q', vim.diagnostic.setloclist, opts)
end

-- Add the server that troubles you here
local name = 'pyright'
local cmd = { 'pyright-langserver', '--stdio' } -- needed for elixirls, omnisharp, sumneko_lua
if not name then
  print 'You have not defined a server name, please edit minimal_init.lua'
end
if not nvim_lsp[name].document_config.default_config.cmd and not cmd then
  print [[You have not defined a server default cmd for a server
    that requires it please edit minimal_init.lua]]
end

nvim_lsp[name].setup {
  cmd = cmd,
  on_attach = on_attach,
}

print [[You can find your log at $HOME/.cache/nvim/lsp.log. Please paste in a github issue under a details tag as described in the issue template.]]
```

### LSP log

https://gist.github.com/tysonliddell/a320c3444fbb6ed0ae3f4de4fd946658

**Comment by tysonliddell** [comment]
Output of `lua =vim.lsp.get_active_clients()[1].server_capabilities` when running pyright. Note that the `workspace` key is not present.
```
{
  callHierarchyProvider = true,
  codeActionProvider = {
    codeActionKinds = { "quickfix", "source.organizeImports" },
    workDoneProgress = true
  },
  completionProvider = {
    completionItem = {
      labelDetailsSupport = true
    },
    resolveProvider = true,
    triggerCharacters = { ".", "[" },
    workDoneProgress = true
  },
  declarationProvider = {
    workDoneProgress = true
  },
  definitionProvider = {
    workDoneProgress = true
  },
  documentHighlightProvider = {
    workDoneProgress = true
  },
  documentSymbolProvider = {
    workDoneProgress = true
  },
  executeCommandProvider = {
    commands = {},
    workDoneProgress = true
  },
  hoverProvider = {
    workDoneProgress = true
  },
  referencesProvider = {
    workDoneProgress = true
  },
  renameProvider = {
    prepareProvider = true,
    workDoneProgress = true
  },
  signatureHelpProvider = {
    triggerCharacters = { "(", ",", ")" },
    workDoneProgress = true
  },
  textDocumentSync = {
    change = 2,
    openClose = true,
    save = {
      includeText = false
    },
    willSave = false,
    willSaveWaitUntil = false
  },
  typeDefinitionProvider = {
    workDoneProgress = true
  },
  workspaceSymbolProvider = {
    workDoneProgress = true
  }
}
```

**Comment by tysonliddell** [comment]
Looks like https://github.com/neovim/nvim-lspconfig/pull/2354 was the right approach. I can confirm that the behaviour is correct with 212b99bc12a5416df8b2a610711ba399e2fc388a, but breaks again with the latest a63d2c75c6880d45209baff596939343e8a431fd.

**Comment by AckslD** [comment]
#2354 also seems to work better than #2356 for me. I often work with one project per tab and when I switch tab with #2354 I still have a single server with the correct root_dir but after #2356 multiple servers are started and the ones for the tab not focused become `single_file_mode`.

I'll try to reproduce with a minimal config and steps later today in a new issue.
